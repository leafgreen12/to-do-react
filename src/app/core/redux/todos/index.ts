export * from './todos.states';
export * from './todos.action';
export * from './todos.reducer';
