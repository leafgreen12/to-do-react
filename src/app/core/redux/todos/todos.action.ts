import {TodoModel} from "../../interfaces/todo.interface";

export enum ETodos {
    ADD_TODO = '[TODOS] add todo',
    EDIT_TODO = '[TODOS] edit todo',
    DELETE_TODO = '[TODOS] delete todo',
}

export class AddTodo {
    public readonly type = ETodos.ADD_TODO;
    constructor(
        public data: TodoModel) {}
}

export class EditTodo {
    public readonly type = ETodos.EDIT_TODO;
    constructor(
        public data: TodoModel) {}
}

export class DeleteTodo {
    public readonly type = ETodos.DELETE_TODO;
    constructor(
        public idTodo: number) {}
}

export type HandleTodoActions =
    AddTodo |
    DeleteTodo |
    EditTodo;
