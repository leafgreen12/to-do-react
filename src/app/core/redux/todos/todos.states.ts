import {TodoModel} from "../../interfaces/todo.interface";

export interface ITodoState {
    items: TodoModel[]
    dispatch?: any
}
