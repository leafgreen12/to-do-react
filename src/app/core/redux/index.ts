import * as fromTodos from './todos';

export interface IAppState {
    todos: fromTodos.ITodoState;
}
