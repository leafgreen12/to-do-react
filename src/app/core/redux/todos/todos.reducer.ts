import {ITodoState} from './todos.states';
import {ETodos, HandleTodoActions} from "./todos.action";
import {TodoModel} from "../../interfaces/todo.interface";

export const initialState = {} as ITodoState

const todos = (state = initialState, action: HandleTodoActions) => {
    switch (action.type) {
        case ETodos.ADD_TODO:
            const newItem = {...state} as ITodoState;
            action.data.id = newItem.items.length + 1;
            newItem.items.push(action.data as TodoModel);
            return newItem;
        case ETodos.EDIT_TODO:
            const newEditItem = {...state} as ITodoState;
            const existItem = newEditItem.items.find(item => item.id === action.data.id);
            if (existItem) {
                existItem.title = action.data.title;
            }
            return newEditItem;
        case ETodos.DELETE_TODO:
            const newArray = {...state} as ITodoState;
            newArray.items = newArray.items.filter(item => item.id !== action.idTodo)
            return newArray;
        default:
            return state
    }
}

export default todos
