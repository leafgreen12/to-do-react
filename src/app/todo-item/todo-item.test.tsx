import {act, render, screen} from '@testing-library/react';
import {unmountComponentAtNode} from 'react-dom';
import Enzyme, {shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TodoItem from "./todo-item";
import {TodoModel} from "../core/interfaces/todo.interface";
import * as fromTodosAction from "../core/redux/todos";

// @ts-ignore
(Enzyme as any).configure({adapter: new Adapter()})
describe('TodoItem', () => {
    let container = null as any;
    let todoItem = null as any;
    beforeEach(() => {
        // cài đặt một DOM element như là target cho render
        container = document.createElement("div");
        document.body.appendChild(container);
        const item = {
            "userId": 1,
            "id": 3,
            "title": "fugiat veniam minus",
            "completed": false
        }
        const editItem = (data: TodoModel) => {};
        todoItem = shallow(<TodoItem item={item} delete={null} edit={editItem}/>)
    });

    afterEach(() => {
        // dọn dẹp lúc thoát
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    it('should render todosLists correctly', () => expect(todoItem).toMatchSnapshot());

    it('toggle edit', () => {
        todoItem.find('.edit-button').simulate('click');
        expect(todoItem.find('.done-button').text()).toBe('Done');
        expect(todoItem.find('.input-text').getElement()).toBeTruthy();
        expect(todoItem).toMatchSnapshot()

        todoItem.find('.done-button').simulate('click');
        expect(todoItem.find('.edit-button').text()).toBe('Edit');

    });
});
