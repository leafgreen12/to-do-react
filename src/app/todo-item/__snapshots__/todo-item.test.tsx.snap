// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App edit enable 1`] = `ShallowWrapper {}`;

exports[`App should render todosLists correctly 1`] = `ShallowWrapper {}`;

exports[`App toggle edit 1`] = `ShallowWrapper {}`;

exports[`App toggle edit 2`] = `ShallowWrapper {}`;

exports[`TodoItem should render todosLists correctly 1`] = `ShallowWrapper {}`;

exports[`TodoItem toggle edit 1`] = `ShallowWrapper {}`;
